### PR TITLE
[#857] Fix some regressions in JPQLNext parser

### LINKED ISSUE #857: Parser regression with top level treat expressions
An entity-view mapping `@Mapping("TREAT(parent AS BaseProduct)")` fails to parse with 1.4.0-Alpha2 that previously worked.

Also, the parser does not restore the original whitespaces in an expression when it throws a SyntaxErrorException.
```
Caused by: com.blazebit.persistence.parser.expression.SyntaxErrorException: A top level treat expression is not allowed. Consider to further dereference the expression: TREAT(parentASBaseProduct)
	at com.blazebit.persistence.parser.expression.JPQLNextExpressionVisitorImpl.visitPathExpression(JPQLNextExpressionVisitorImpl.java:599)
	at com.blazebit.persistence.parser.expression.JPQLNextExpressionVisitorImpl.visitPathExpression(JPQLNextExpressionVisitorImpl.java:56)
	at com.blazebit.persistence.parser.JPQLNextParser$PathExpressionContext.accept(JPQLNextParser.java:6638)
	at com.blazebit.persistence.parser.expression.JPQLNextExpressionVisitorImpl.visitParseExpression(JPQLNextExpressionVisitorImpl.java:94)
	at com.blazebit.persistence.parser.expression.JPQLNextExpressionVisitorImpl.visitParseExpression(JPQLNextExpressionVisitorImpl.java:56)
	at com.blazebit.persistence.parser.JPQLNextParser$ParseExpressionContext.accept(JPQLNextParser.java:408)
	at com.blazebit.persistence.parser.antlr.tree.AbstractParseTreeVisitor.visit(AbstractParseTreeVisitor.java:18)
	at com.blazebit.persistence.parser.expression.AbstractExpressionFactory.createExpression(AbstractExpressionFactory.java:186)
	at com.blazebit.persistence.parser.expression.AbstractExpressionFactory.createSimpleExpression(AbstractExpressionFactory.java:213)
	at com.blazebit.persistence.parser.expression.AbstractCachingExpressionFactory$2.get(AbstractCachingExpressionFactory.java:56)
	at com.blazebit.persistence.parser.expression.AbstractCachingExpressionFactory.getOrDefault(AbstractCachingExpressionFactory.java:158)
	at com.blazebit.persistence.parser.expression.AbstractCachingExpressionFactory.createSimpleExpression(AbstractCachingExpressionFactory.java:109)
	at com.blazebit.persistence.view.impl.MacroConfigurationExpressionFactory.createSimpleExpression(MacroConfigurationExpressionFactory.java:67)
	at com.blazebit.persistence.parser.expression.AbstractExpressionFactoryMacroAdapter.createSimpleExpression(AbstractExpressionFactoryMacroAdapter.java:66)
	at com.blazebit.persistence.view.impl.metamodel.AbstractAttribute.getCollectionJoinMappings(AbstractAttribute.java:395)
	at com.blazebit.persistence.view.impl.metamodel.ManagedViewTypeImpl.checkAttributes(ManagedViewTypeImpl.java:391)
	at com.blazebit.persistence.view.impl.metamodel.ViewMetamodelImpl.<init>(ViewMetamodelImpl.java:103)
	at com.blazebit.persistence.view.impl.EntityViewManagerImpl.<init>(EntityViewManagerImpl.java:166)
```

### PR DESCRIPTION
**Pull request by Mobe91**
Fixes #857

### CODE REVIEW

File: core/parser/src/main/java/com/blazebit/persistence/parser/expression/JPQLNextExpressionVisitorImpl.java [hunk 1]
```diff
@@ -1026,4 +1030,11 @@ private Expression createEntityTypeLiteral(String entityLiteralStr) {
         }
         return new EntityLiteral(entityType, entityLiteralStr);
     }
+
+    private String getInputText(ParserRuleContext ctx) {
+        int from = ctx.start.getStartIndex();
+        int to = ctx.stop.getStopIndex();
+        Interval interval = new Interval(from,to);
```
Comment (beikov) [inline core/parser/src/main/java/com/blazebit/persistence/parser/expression/JPQLNextExpressionVisitorImpl.java]:
```suggestion
        Interval interval = new Interval(from, to);
```

**Comment by Mobe91** [comment]
Not really, the cases that you mention did not change compared to before your parser rework commit so I did not touch them.
I would need to disable expression caching to create a test case that triggers this behavior - ok?

**Comment by beikov** [comment]
True but the cases I mentioned should also allow object results.
You shouldn't need a cache to reproduce this. If you do, that would be a bug.

**Comment by Mobe91** [comment]
> True but the cases I mentioned should also allow object results.

Ok, I will apply the changes then.

> You shouldn't need a cache to reproduce this. If you do, that would be a bug.

Expression caching itself is a bug at the moment. The cache keys need to consider the various flags that influence behavior of parsing and the expression visitor which is currently not the case. If it was implemented that way, those regressions (and possibly the additional cases you mentioned) would have come up with the existing tests I think.

I can fix the expression caching in the course of this issue.

**Comment by beikov** [comment]
Please create constants for the bit masks

**Comment by beikov** [comment]
Checkstyle issues: https://travis-ci.org/Blazebit/blaze-persistence/jobs/592945474#L953

File: core/parser/src/main/java/com/blazebit/persistence/parser/expression/JPQLNextExpressionVisitorImpl.java (hunk 1 above)
Comment (beikov) [reply core/parser/src/main/java/com/blazebit/persistence/parser/expression/JPQLNextExpressionVisitorImpl.java]: Please fix that and squash the commits

**Pull request merged**